drivers/media/pci/intel/ipu3/cio2-bridge.c

```diff
@@ -72,11 +73,36 @@ static int cio2_bridge_read_acpi_buffer(struct acpi_device *adev, char *id,
 	return ret;
 }
 
+static u32 cio2_bridge_parse_rotation(u8 rotation)
+{
+	if (rotation == 1)
+		return 180;
+	return 0;
+}
+
+static enum v4l2_fwnode_orientation cio2_bridge_parse_orientation(u8 panel)
+{
+	switch (panel) {
+	case 4:
+		return V4L2_FWNODE_ORIENTATION_FRONT;
+	case 5:
+		return V4L2_FWNODE_ORIENTATION_BACK;
+	default:
+		return V4L2_FWNODE_ORIENTATION_EXTERNAL;
+	}
+}
+
```

Comment:
We don't have any spec for this, right? Maybe print a warning or debug for any unknown value?

Reply:
The rotation is reverse engineered so 1 == 180 is just a guess but a comment is indeed necessary.

For the orientation we have [this spec](https://uefi.org/sites/default/files/resources/ACPI_6_3_final_Jan30.pdf) on page 351 so I'll add a comment for that as well.

What do you think about the switch-statement? Is there a better way of doing the conversion?

Reply:
I think you'd probably find / need to add macros for the 4/5 values for the _PLD panel field when this goes upstream. The switch itself is fine imo.

I'm surprised to find 1==180, as it sort of precludes a sensor being mounted at 90 degrees or something...I was expecting something like  1 = 90 degrees, 2 = 180. Anyway, I think that probably ought to return the input value in the case that it's not 1, otherwise if we do find some other value in there it'll be masked here.

Reply:
> For the orientation we have [this spec](https://uefi.org/sites/default/files/resources/ACPI_6_3_final_Jan30.pdf) on page 351 so I'll add a comment for that as well.

Thanks for the link! Maybe mention that this value comes from the ACPI PLD buffer, then people can find the spec for that themselves. Also since anything except front and back seem to be directly translatable from ACPI, I think you should go with a warning here (as opposed to a debug print).

> I think you'd probably find / need to add macros for the 4/5 values for the _PLD panel field when this goes upstream. The switch itself is fine imo.

Yeah, I think those should be constants if they're defined in ACPI. Maybe also for the rotation.

> What do you think about the switch-statement? Is there a better way of doing the conversion?

I agree with @djrscally. I think switch is probably the best option here. A lookup-table would be an alternative, but doesn't make much sense here, I think.

> I'm surprised to find 1==180, as it sort of precludes a sensor being mounted at 90 degrees or something...I was expecting something like 1 = 90 degrees, 2 = 180.

I assume we don't know any device that has a 90 degree sensor so we could check this?

> Anyway, I think that probably ought to return the input value in the case that it's not 1, otherwise if we do find some other value in there it'll be masked here.

I'm not entirely sure if returning the input value is a good thing, but there should be something to help us figure out if there's an unknown value (that's why I proposed warning or debug messages).

Reply:
> Maybe also for the rotation.

Yeah, they could just got in cio2-bridge.h

> I assume we don't know any device that has a 90 degree sensor so we could check this?

Not to my knowledge, I don't recall anything except a 180 rotation, I was just going off what Windows apparently expects from the rotation field in _PLD: https://docs.microsoft.com/en-us/windows-hardware/drivers/stream/camera-device-orientation.

Then again, given the field is called "degree" you'd really expect it to be in degrees if anything!

> I'm not entirely sure if returning the input value is a good thing, but there should be something to help us figure out if there's an unknown value (that's why I proposed warning or debug messages).

Fair enough; a debug print is good for me too then

Reply:
> Then again, given the field is called "degree" you'd really expect it to be in degrees if anything!

Yeah... let's just hope the 0/1 thing is consistent for now and we don't end up with other devices actually having a rotation in degrees there. Might be a bit messy handling that.

Reply:
It's also strange that `degree` is only 8-bit so it cannot encode e.g. 270 degree directly.

I'll try to go through the SSDB buffers and find all values for degree then we have a better picture.

Reply:
I grepped the ACPI info from [here](https://github.com/kitakar5525/surface-ipu3-cameras/tree/master/misc/dump_intel_ipu_data/results) and most of the sensors have 0 or 1 in the degree field (SSDB) and 0 in the roation field (PLD).

There are some quirks though:
- Switch_SA5-271 has the degree field set to 180 (yey!)
- Surface_Book_3 and Surface_Pro_7 report 2 (i.e. 90 degree) in the PLD roation field for the IR-Sensor

I think it's save to treat degree as a "binary" variable for now and treat the three devices separately.

Reply:
> There are some quirks though:
> 
> * Switch_SA5-271 has the degree field set to 180 (yey!)

Whelp, of course there are... would be too easy otherwise, would it. I think we can assume though that it's either 0, 1, 2, (maybe 3?) and 90, 180. So still possible to somewhat sanely handle those.

>  * Surface_Book_3 and Surface_Pro_7 report 2 (i.e. 90 degree) in the PLD roation field for the IR-Sensor

Hmm, too bad those are IPU4 devices. But at least we now know that we can maybe check these values in the future.

Reply:
> There are some quirks though:
>
>    Switch_SA5-271 has the degree field set to 180 (yey!)

Oh good grief! 

> I think it's save to treat degree as a "binary" variable for now and treat the three devices separately.

Yeah, sounds sensible to me.